File: java/src/main/java/ai/pagou/examples/lib/Redactor.java
package ai.pagou.examples.lib;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Masks sensitive values before anything is logged so secrets, tokens and card
 * data never reach stdout or a log sink.
 */
public final class Redactor {

  private static final Set<String> SENSITIVE_KEYS =
      Set.of(
          "authorization",
          "apikey",
          "api_key",
          "token",
          "access_token",
          "client_secret",
          "secret",
          "password",
          "cvv",
          "cvc",
          "pan",
          "card_number",
          "number");

  private static final Pattern[] TOKEN_PATTERNS = {
    Pattern.compile("\\bBearer\\s+[A-Za-z0-9._-]+", Pattern.CASE_INSENSITIVE),
    Pattern.compile("\\bpg(ct|pm|sk|pk)_[A-Za-z0-9]+"),
  };

  private static final String REDACTED = "[REDACTED]";

  private Redactor() {}

  /** Masks bearer tokens and {@code pg*_} tokens embedded in a free-text string. */
  public static String redactString(String value) {
    String out = value;
    for (Pattern pattern : TOKEN_PATTERNS) {
      out = pattern.matcher(out).replaceAll(REDACTED);
    }
    return out;
  }

  /** Deep-copies a JSON tree with sensitive keys masked and tokens scrubbed. */
  public static JsonNode redact(JsonNode node) {
    if (node == null || node.isNull()) {
      return node;
    }
    if (node.isTextual()) {
      return Json.MAPPER.getNodeFactory().textNode(redactString(node.asText()));
    }
    if (node.isArray()) {
      ArrayNode out = Json.MAPPER.createArrayNode();
      for (JsonNode item : node) {
        out.add(redact(item));
      }
      return out;
    }
    if (node.isObject()) {
      ObjectNode out = Json.MAPPER.createObjectNode();
      Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
      while (fields.hasNext()) {
        Map.Entry<String, JsonNode> entry = fields.next();
        if (SENSITIVE_KEYS.contains(entry.getKey().toLowerCase())) {
          out.put(entry.getKey(), REDACTED);
        } else {
          out.set(entry.getKey(), redact(entry.getValue()));
        }
      }
      return out;
    }
    return node;
  }

  /** Convenience: redact an arbitrary object by round-tripping through JSON. */
  public static JsonNode redactValue(Object value) {
    return redact(Json.MAPPER.valueToTree(value));
  }
}

File: dotnet/src/Pagou.Examples.Core/Redactor.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace Pagou.Examples.Core;

/// <summary>
/// Masks sensitive fields before anything is logged so secrets, tokens and card
/// data never reach stdout or a log sink.
/// </summary>
public static partial class Redactor
{
    private static readonly HashSet<string> SensitiveKeys = new(StringComparer.OrdinalIgnoreCase)
    {
        "authorization", "apikey", "api_key", "token", "access_token", "client_secret",
        "secret", "password", "cvv", "cvc", "pan", "card_number", "number",
    };

    private const string RedactedText = "[REDACTED]";

    [GeneratedRegex(@"\bBearer\s+[A-Za-z0-9._-]+", RegexOptions.IgnoreCase)]
    private static partial Regex BearerPattern();

    [GeneratedRegex(@"\bpg(ct|pm|sk|pk)_[A-Za-z0-9]+")]
    private static partial Regex CardTokenPattern();

    /// <summary>Masks bearer tokens and card tokens embedded in free text.</summary>
    public static string RedactText(string value)
    {
        value = BearerPattern().Replace(value, RedactedText);
        value = CardTokenPattern().Replace(value, RedactedText);
        return value;
    }

    /// <summary>Serializes a value to JSON and returns a redacted copy of it.</summary>
    public static JsonNode? RedactToNode(object? value)
    {
        var node = value as JsonNode ?? (value is null ? null : JsonSerializer.SerializeToNode(value, Json.Options));
        return Redact(node);
    }

    /// <summary>Recursively rebuilds a node with sensitive keys and token strings masked.</summary>
    public static JsonNode? Redact(JsonNode? node)
    {
        switch (node)
        {
            case JsonObject obj:
                var result = new JsonObject();
                foreach (var pair in obj)
                {
                    result[pair.Key] = SensitiveKeys.Contains(pair.Key)
                        ? RedactedText
                        : Redact(pair.Value);
                }
                return result;

            case JsonArray array:
                var redactedArray = new JsonArray();
                foreach (var item in array)
                {
                    redactedArray.Add(Redact(item));
                }
                return redactedArray;

            case JsonValue value:
                return value.TryGetValue<string>(out var text)
                    ? JsonValue.Create(RedactText(text))
                    : value.DeepClone();

            default:
                return null;
        }
    }
}
